Add support for mypy

Like skeleton projects, I'd like to perform type checks on the code base as a standard practice.

Probably the process is similar to how ruff is invoked (grab the upstream config, save it in a context, and then run the tests with `pytest-mypy` installed).

On further consideration, the warning issued above is present when there's no mypy config, but when adding the mypy config from skeleton, the warning goes away and an error is manifest instead.

```
______________________________________________________________________ [mypy] __main__.py _______________________________________________________________________
__main__.py: error: Duplicate module named "__main__" (also at "__init__.py")
__main__.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
__main__.py: note: Common resolutions include:
__main__.py: note:     a) using `--exclude` to avoid checking one of them,
__main__.py: note:     b) adding `__init__.py` somewhere,
__main__.py: note:     c) using `--explicit-package-bases` or adjusting `MYPYPATH`
============================================================================= mypy ==============================================================================
Found 1 error in 1 file (errors prevented further checking)
==================================================================== short test summary info ====================================================================
FAILED __main__.py::mypy - __main__.py: error: Duplicate module named "__main__" (also at "__init__.py")
```

With c48ee66, I figured out a way to patch the mypy discovery so it honors the essential layout. I'm not confident it will work for submodules; we shall see.

Now it's only emitting errors that are seemingly more specific to the coherent.test project (and its dependencies).

```
_______________________________________________________________________ [mypy] core.py _________________________________________________________________________
core.py:10: error: Skipping analyzing "pip_run.deps": module is installed, but missing library stubs or py.typed marker  [import-untyped]
core.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
core.py:10: error: Skipping analyzing "pip_run": module is installed, but missing library stubs or py.typed marker  [import-untyped]
core.py:11: error: Skipping analyzing "pip_run.launch": module is installed, but missing library stubs or py.typed marker  [import-untyped]
core.py:12: error: Skipping analyzing "coherent.build": module is installed, but missing library stubs or py.typed marker  [import-untyped]
core.py:57: error: Argument 1 to "contextmanager" has incompatible type "Callable[[], None]"; expected "Callable[[], Iterator[Never]]"  [arg-type]
core.py:58: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
_______________________________________________________________________ [mypy] monkey.py ________________________________________________________________________
monkey.py:7: error: Skipping analyzing "coherent.build": module is installed, but missing library stubs or py.typed marker  [import-untyped]
monkey.py:7: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
monkey.py:32: error: Cannot find implementation or library stub for module named "mypy.find_sources"  [import-not-found]
monkey.py:32: error: Cannot find implementation or library stub for module named "mypy"  [import-not-found]
monkey.py:40: error: "super" used outside class  [misc]
```

--- a/__init__.py
+++ b/__init__.py
@@ -9,5 +9,6 @@ __requires__ = [
     'pytest',
     'pytest-enabler',
     'pytest-ruff',
-    'pytest-mypy',
+    # Exclude PyPy from type checks (python/mypy#20454 jaraco/skeleton#187)
+    "pytest-mypy; platform_python_implementation != 'PyPy'",
 ]

--- a/__init__.py
+++ b/__init__.py
@@ -9,4 +9,5 @@ __requires__ = [
     'pytest',
     'pytest-enabler',
     'pytest-ruff',
+    'pytest-mypy',
 ]

--- a/core.py
+++ b/core.py
@@ -33,21 +33,25 @@ def build_env(target, *, orig=os.environ):
     return {**orig, **overlay}
 
 
-def load_ruff_toml():
-    url = 'https://raw.githubusercontent.com/jaraco/skeleton/refs/heads/main/ruff.toml'
+def load_skeleton_file(name):
+    url = f'https://raw.githubusercontent.com/jaraco/skeleton/refs/heads/main/{name}'
     return urllib.request.urlopen(url).read().decode('utf-8')
 
 
-def configure_ruff():
+def configure(name):
     """
     >>> getfixture('monkeypatch').chdir(getfixture('tmp_path'))
-    >>> with configure_ruff():
+    >>> with configure('ruff.toml'):
     ...     pathlib.Path('ruff.toml').stat().st_size > 0
     True
     """
-    if pathlib.Path('(meta)/ruff.toml').exists():
+    if pathlib.Path(f'(meta)/{name}').exists():
+        # for future, consider honoring this file
         raise NotImplementedError
-    return bootstrap.assured(pathlib.Path('ruff.toml'), load_ruff_toml)
+    return bootstrap.assured(
+        pathlib.Path(name),
+        functools.partial(load_skeleton_file, name),
+    )
 
 
 @contextlib.contextmanager
@@ -56,7 +60,12 @@ def project_on_path():
     Install the project under test and yield its new install path.
     """
     deps = pip_run.deps.load('--editable', '.[test]')
-    with bootstrap.write_pyproject(), deps as home, configure_ruff():
+    with (
+        bootstrap.write_pyproject(),
+        deps as home,
+        configure('ruff.toml'),
+        configure('mypy.ini'),
+    ):
         yield home
 
 

--- a/core.py
+++ b/core.py
@@ -7,9 +7,9 @@ import sys
 import urllib.request
 
 import jaraco.functools
-import pip_run.deps
-import pip_run.launch
-from coherent.build import bootstrap
+import pip_run.deps  # type: ignore[import-untyped]
+import pip_run.launch  # type: ignore[import-untyped]
+from coherent.build import bootstrap  # type: ignore[import-untyped]
 
 
 def build_env(target, *, orig=os.environ):

--- a/monkey.py
+++ b/monkey.py
@@ -6,6 +6,7 @@ from typing import TYPE_CHECKING
 
 from coherent.build import discovery  # type: ignore[import-untyped]
 from jaraco.compat.py38 import cache, r_fix
+from jaraco.context import suppress
 
 if TYPE_CHECKING:
     from types import ModuleType
@@ -28,6 +29,7 @@ def import_path(path: StrPath, *, root: pathlib.Path, **unused_kwargs) -> Module
     )
 
 
+@suppress(ImportError)
 def patch_mypy():
     import mypy.find_sources  # type: ignore[import-not-found]
 

--- a/monkey.py
+++ b/monkey.py
@@ -4,7 +4,7 @@ import importlib
 import pathlib
 from typing import TYPE_CHECKING
 
-from coherent.build import discovery
+from coherent.build import discovery  # type: ignore[import-untyped]
 from jaraco.compat.py38 import cache, r_fix
 
 if TYPE_CHECKING:
@@ -29,7 +29,7 @@ def import_path(path: StrPath, *, root: pathlib.Path, **unused_kwargs) -> Module
 
 
 def patch_mypy():
-    import mypy.find_sources
+    import mypy.find_sources  # type: ignore[import-not-found]
 
     class EssentialFinder(mypy.find_sources.SourceFinder):
         def crawl_up_dir(self, dir: str) -> tuple[str, str]:

--- a/monkey.py
+++ b/monkey.py
@@ -28,12 +28,37 @@ def import_path(path: StrPath, *, root: pathlib.Path, **unused_kwargs) -> Module
     )
 
 
-def patch_all():
+def patch_mypy():
+    import mypy.find_sources
+
+    def crawl_up_dir(self, dir: str) -> tuple[str, str]:
+        """
+        Honor essential layout in mypy SourceFinder.
+        """
+        if pathlib.Path().samefile(dir):
+            return best_name(), dir
+        return super().crawl_up_dir(dir)
+
+    EssentialFinder = type(
+        'EssentialFinder',
+        (mypy.find_sources.SourceFinder,),
+        dict(crawl_up_dir=crawl_up_dir),
+    )
+
+    mypy.find_sources.SourceFinder = EssentialFinder
+
+
+def patch_pytest():
     import _pytest.config
     import _pytest.python
 
     _pytest.config.import_path = import_path
     _pytest.python.import_path = import_path
+
+
+def patch_all():
+    patch_pytest()
+    patch_mypy()
 
 
 def pytest_configure():

--- a/monkey.py
+++ b/monkey.py
@@ -31,19 +31,14 @@ def import_path(path: StrPath, *, root: pathlib.Path, **unused_kwargs) -> Module
 def patch_mypy():
     import mypy.find_sources
 
-    def crawl_up_dir(self, dir: str) -> tuple[str, str]:
-        """
-        Honor essential layout in mypy SourceFinder.
-        """
-        if pathlib.Path().samefile(dir):
-            return best_name(), dir
-        return super().crawl_up_dir(dir)
-
-    EssentialFinder = type(
-        'EssentialFinder',
-        (mypy.find_sources.SourceFinder,),
-        dict(crawl_up_dir=crawl_up_dir),
-    )
+    class EssentialFinder(mypy.find_sources.SourceFinder):
+        def crawl_up_dir(self, dir: str) -> tuple[str, str]:
+            """
+            Honor essential layout in mypy SourceFinder.
+            """
+            if pathlib.Path().samefile(dir):
+                return best_name(), dir
+            return super().crawl_up_dir(dir)
 
     mypy.find_sources.SourceFinder = EssentialFinder
 

